### PR TITLE
Valgrind memcheck api integration in GC

### PR DESCRIPTION
**Pull request by Varun-garg**
I would like to share work that I and @youngar  have done to add Valgrind Memcheck support in garbage collector. This covers my submission for Google Summer of Code.

**Proposal**: https://summerofcode.withgoogle.com/serve/6741589379514368/
**Mentor**: Andrew Young

#### Work done
1. Memcheck has been integrated with exampleVM. `omrgcexample`, runs fine.
2. All garbage collection schemes have also been integrated with memcheck. All tests (8 for gc and 2 for performance) run successfully. 
3. All calls are made from inside `/gc` directory. There is no need to re-implement them in target language.
4. All Valgrind requests are made through a wrapper file:  https://github.com/Varun-garg/omr/blob/59db4760c478c8415450e1d4746606fc91854bd8/gc/base/MemcheckWrapper.hpp . It is made to increase abstraction and also to replace existing methods with better APIs in future.
5. **Logging:** extensive valgrind logs can be generated by enabling `VALGRIND_REQUEST_LOGS` in wrapper file.
6. **Light testing:** While destroying memory base, an assertion has been placed to check if all objects allocated through API were freed using it.

#### Building OMR with Valgrind API
To compile with valgrind API enabled, use build option  `--enable-OMR_VALGRIND_MEMCHECK` while generating makefiles. Valgrind documentation also recommends turning off optimizations.

#### Running GC on Valgrind (and vgdb) and debugging.
This section is covered in documentation (link below).

#### Documentation 
https://github.com/Varun-garg/omr/blob/59db4760c478c8415450e1d4746606fc91854bd8/doc/Valgrind%20Memcheck%20API.md

#### Commits
33 Commits (+891 −261) have been pushed by us in GSoC session. Last commit is: 91e1cfa05de6b757a0730949d4036f6f302a969a

#### Future plans
These are some things that I am planning to do in future.
1. Run `VALGRIND_CHECK_MEM_IS_DEFINED` while freeing objects. 
2. Run Valgrind on Travis.
3. Implement redzones while creating objects.

Issue: #826 
Signed-off-by: Varun Garg <varun.10@live.com>

### CODE REVIEW

File: gc/base/GCExtensionsBase.hpp [hunk 1]
```diff
@@ -723,6 +723,8 @@ class MM_GCExtensionsBase : public MM_BaseVirtual {
 	bool compactOnIdle; /**< Forces compaction if global GC executed while VM Runtime State set to IDLE, default is false */
 #endif
 
+	uintptr_t valgrindMemppolAddr; /** <Memory pool's address for valgrind> **/
```
Comment (youngar) [inline gc/base/GCExtensionsBase.hpp]: This should be wrapped in `#if defined(OMR_VALGRIND_MEMCHECK)`

File: gc/base/MemoryManager.cpp [hunk 2]
```diff
@@ -68,8 +71,8 @@ MM_MemoryManager::createVirtualMemoryForHeap(MM_EnvironmentBase* env, MM_MemoryH
 
 	uintptr_t allocateSize = size;
 
-	uintptr_t concurrentScavengerPageSize = 0;
 	uintptr_t concurrentScavengerPageAlignmentIncrement = 0;
+	uintptr_t concurrentScavengerPageSize = 0;
```
Comment (youngar) [inline gc/base/MemoryManager.cpp]: This looks like an accidental line switch.

**Comment by Varun-garg** [comment]
@youngar Thank you for reviewing this request. I have updated code according to it.

**Comment by lmaisons** [comment]
Can some of these commits get squashed?  Some of them look like they wouldn't be meaningful post-development.  For instance, if 'fixed issue with sizeof' isn't fixing a defect that already exists in master, it would be a primary candidate for squashing.

**Comment by Varun-garg** [comment]
@lmaisons Thank you for going through the PR. I have squashed 8 commits now.

**Comment by Varun-garg** [comment]
Valgrind now works with 4/8 gc tests:

1. fvtest/gctest/configuration/sample_GC_config.xml
2. fvtest/gctest/configuration/test_system_gc.xml
3. fvtest/gctest/configuration/global_GC_config.xml
4. fvtest/gctest/configuration/optavgpause_GC_config.xml

**Comment by Varun-garg** [comment]
Valgrind free object calls are now moved from exampleVM to GC. There are only few objects throwing errors that need to be resolved.

**Comment by mgaudet** [comment]
Regarding the Future Plans, would it be possible to get followup issues opened?

**Comment by Varun-garg** [comment]
Sure @mgaudet , actually I am currently stuck with mid semester exams, in a few more days I will resume the project.

**Comment by mgaudet** [comment]
No rush :) Just want follow up issues not to get lost, as well, if they're open, others may be able to help.

**Comment by Varun-garg** [comment]
Hey @mgaudet , I am now free so we can work solve followup issues. API improvements are continued in #1759

**Comment by mgaudet** [comment]
Anything you're not tackling actively, what I'd suggest is opening an Issue. This just makes it easier for others to jump on and help should they want to.

**Comment by Varun-garg** [comment]
@mgaudet Thank you for the advice. I will open an issue shortly for a couple of queries.

**Comment by Varun-garg** [comment]
@mgaudet I have opened issue #1799 . It has some points that can make Valgrind integration better.

**Comment by mgaudet** [comment]
Fantastic. Thanks so much!

**Pull request merged**